source/particle/generator/random_uniform.cc

```diff
@@ -0,0 +1,205 @@
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file doc/COPYING.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/generator/random_uniform.h>
+
+#include <boost/random.hpp>
+
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Generator
+    {
+      template <int dim>
+      RandomUniform<dim>::RandomUniform() {}
+
+      template <int dim>
+      std::multimap<types::LevelInd, Particle<dim> >
+      RandomUniform<dim>::generate_particles()
+      {
+        // Calculate the number of particles in this domain as a fraction of total volume
+        double local_volume = 0;
+        for (typename parallel::distributed::Triangulation<dim>::active_cell_iterator
+             it=this->get_triangulation().begin_active();
+             it!=this->get_triangulation().end(); ++it)
+          {
+            const double cell_volume = it->measure();
+            AssertThrow (cell_volume != 0, ExcMessage ("Found cell with zero volume."));
+
+            if (it->is_locally_owned())
+              local_volume += cell_volume;
+          }
+
+        // Sum the local volumes over all nodes
+        const double total_volume = Utilities::MPI::sum(local_volume,this->get_mpi_communicator());
+
+        // Assign this subdomain the appropriate fraction
+        double subdomain_fraction = local_volume / total_volume;
+
+        // Sum the subdomain fractions so we don't miss particles from rounding and to create unique IDs
+        double end_fraction = 0.0;
+        MPI_Scan(&subdomain_fraction, &end_fraction, 1, MPI_DOUBLE, MPI_SUM, this->get_mpi_communicator());
+        const double start_fraction = end_fraction - subdomain_fraction;
+
+        // Calculate start and end IDs so there are no gaps
+        // TODO: this can create gaps for certain processor counts because of
+        // floating point imprecision, figure out how to fix it
+        const types::particle_index start_id = static_cast<types::particle_index>(std::ceil(start_fraction*n_tracers));
+        const types::particle_index end_id   = static_cast<types::particle_index>(fmin(std::ceil(end_fraction*n_tracers), n_tracers));
+        const types::particle_index subdomain_particles = end_id - start_id;
+
+        return uniform_random_particles_in_subdomain(subdomain_particles, start_id);
+      }
+
+      template <int dim>
+      std::multimap<types::LevelInd, Particle<dim> >
+      RandomUniform<dim>::uniform_random_particles_in_subdomain (const types::particle_index subdomain_particles,
+                                                                 const types::particle_index start_id)
```

Comment:
maybe call the function `create_uniform_...` to indicate its intent? I think the same applied to the other classes.